### ISSUE TITLE
make_customize_airootfs is first before setup AUR or PKGBUILDS.

### ISSUE THREAD

**Issue by sapphire1896** [FascodeNet/alterlinux#283]
Must config newfs‘s sudoers is all.

Correct
run_once make_packages_repo
run_once make_customize_airootfs
[[ "${noaur}" = false ]] && run_once make_packages_aur
[[ "${nopkgbuild}" = false ]] && run_once make_pkgbuild

Refer archlinux/archiso/mkarchiso
    [[ "${quiet}" == "y" ]] || _show_config
    _run_once _make_pacman_conf
    [[ -z "${gpg_key}" ]] || _run_once _export_gpg_publickey
    _run_once _make_custom_airootfs
    _run_once _make_packages
    _run_once _make_version
    _run_once _make_customize_airootfs
    _run_once _make_pkglist

Notice
To run make_packages_repo is first, we could correctly read special device it will be need on the chroot environment. Working ideal by /etc/sudoers.d; automatic confirm, init keyrings and resolve DNS with arch-chroot.

No needed we have to rewrite build.sh anymore ultimately.

If we had to face, partially move airootfs.sh has some of the basis into function.sh, is the simplest correction.

**Comment by Hayao0819** [comment]
It is not my mistakes and it came from a historical reason.

The part of current AlterISO 3.1 and AlterISO 4 are based on Archiso v43.

At that time, Archiso had only function named `make_customize_airootfs` , which run after `make_packages`.

That function did 2 things, which are "Copy files to airootfs" and "run customize_airootfs.sh".

However, Archiso has changed the struct of codes since Archiso v47.

You can see the comparing of them [here](https://github.com/archlinux/archiso/compare/v46...v47)

The important things is one change that change the order of function, which run file copying before installing as you know.

Normally, pacman says error if it try to overwrite files, but there is one exception, the package has `backup` in PKGBUILD.

Archiso only changes minimal config files so the change is not issue. Archiso maintainer said that the way of the change is for secure. so it is not important.

But AlterISO will try to overwrite a lot of files even if it is not included in `backup` array, and the reason is simply, to customize a lot for Japanese user.

If I have many time, It is not impossible to re-write all channels for new function order or even re-write all channels for Archiso-config structure.



In addition all `customize_airootfs.sh` will expect that all aur packages has been installed.

I have a lot of channels which is not included in official repository and I emphasize compatibility than Archiso.

In summary, the reason of the order of code is "lack of times" and "history"

**Comment by Hayao0819** [comment]
Also, I cannot understand your issue after `Notice`

**Comment by sapphire1896** [comment]
結果的に誤報告でした、すみませんm_m

必要な一部の関数を移動させた時に場所が違ったので、この順番はあってまして歴史的な経緯もその通りです。

別の変更も試してしまっているため、もし修正の提案があれば落ち着いてこちらに改めて打診したいと思います。(今のところpkglist.shやGPG関連を読んでいました)

**Comment by Hayao0819** [comment]
👍
